lib/rubocop/cop/performance/reverse_each.rb

```diff
@@ -3,44 +3,59 @@
 module RuboCop
   module Cop
     module Performance
-      # This cop is used to identify usages of `reverse.each` and
-      # change them to use `reverse_each` instead.
+      # This cop is used to identify usages of `reverse.each` (or
+      # `reverse.each_with_index`) and change them to use
+      # `reverse_each` (or `reverse_each.with_index`) instead.
       #
       # @example
       #   # bad
       #   [].reverse.each
+      #   [].reverse.each_with_index
       #
       #   # good
       #   [].reverse_each
+      #   [].reverse_each.with_index
       class ReverseEach < Base
         include RangeHelp
         extend AutoCorrector
 
-        MSG = 'Use `reverse_each` instead of `reverse.each`.'
-        RESTRICT_ON_SEND = %i[each].freeze
-        UNDERSCORE = '_'
+        MSG_EACH = 'Use `reverse_each` instead of `reverse.each`.'
+        MSG_EACH_WITH_INDEX = 'Use `reverse_each.with_index` instead of `reverse.each_with_index`.'
+        RESTRICT_ON_SEND = %i[each each_with_index].freeze
 
         def_node_matcher :reverse_each?, <<~MATCHER
           (send $(send _ :reverse) :each)
```

Comment:
I think `{...}` can be used here to simplify the pattern and code.

```diff
diff --git a/lib/rubocop/cop/performance/reverse_each.rb b/lib/rubocop/cop/performance/reverse_each.rb
index 17629f510..48d8c9e68 100644
--- a/lib/rubocop/cop/performance/reverse_each.rb
+++ b/lib/rubocop/cop/performance/reverse_each.rb
@@ -19,31 +19,26 @@ module RuboCop
         include RangeHelp
         extend AutoCorrector

-        MSG_EACH = 'Use `reverse_each` instead of `reverse.each`.'
-        MSG_EACH_WITH_INDEX = 'Use `reverse_each.with_index` instead of `reverse.each_with_index`.'
+        MSG = 'Use `reverse%<prefer_method>s` instead of `reverse.%<current_method>s`.'
         RESTRICT_ON_SEND = %i[each each_with_index].freeze

         def_node_matcher :reverse_each?, <<~MATCHER
-          (send $(send _ :reverse) :each)
-        MATCHER
-
-        def_node_matcher :reverse_each_with_index?, <<~MATCHER
-          (send $(send _ :reverse) :each_with_index)
+          (send $(send _ :reverse) ${:each :each_with_index})
         MATCHER

         def on_send(node)
-          reverse_each?(node) do |receiver|
-            handle_offense(receiver, node, MSG_EACH, '_each')
-          end
+          reverse_each?(node) do |receiver, current_method|
+            prefer_method = current_method == :each ? '_each' : '_each.with_index'
+            message = format(MSG, current_method: current_method, prefer_method: prefer_method)

-          reverse_each_with_index?(node) do |receiver|
-            handle_offense(receiver, node, MSG_EACH_WITH_INDEX, '_each.with_index')
+            register_offense(receiver, node, message, prefer_method)
           end
         end

         private

-        def handle_offense(receiver, node, msg, replacement)
+        def register_offense(receiver, node, msg, replacement)
           location_of_reverse = receiver.loc.selector.begin_pos
           end_location = node.loc.selector.end_pos
```